# **HW02 - EDA cycle.**

## Задача: найти датасет *(много данных, много категорий, есть пропуски)* и провести EDA.
- Сбор.
- Очистка.
- Исследование.
- Предобработка.
- Гипотезы и выводы.

## 1) Выбор датасета
В качестве датасета я выбрал **Spotify Tracks Dataset** с kaggle *(https://www.kaggle.com/datasets/maharshipandya/-spotify-tracks-dataset)*
### Пару слов о датасете (кроме очевидных параметров):
- explicit - Откровенные тексты (я не знаю как это нормально и правильно перевести).
- danceability - от 0 до 1 насколько песня танцевальная.
- energy - здесь понимается скорость, громкость, энергия песни. од 0 до 1 (спокойная, энергичная).
- key - тональность песни. 0 - C, 1 - C#, и т.д.
- loudness - громкость песни в децибеллах.
- mode - минорная (0) или мажорная (1) тональность.
- speechiness - показывает как много текста в песне. от 0 до 1. 1 - подкаст, 0 - мелодия, выше 0.66 - песня где много слов, 0.33 - 0.66, как правило обычная песня, например рэп, ниже 0.33 - песня, где мало слов.
- acousticness - показывает вероятность, с которой песня аккустическая.
- instrumentalness - показывает вероятность, с которой песня почти без вокала.
- liveness - показывает вероятность, с которой песня была записана в лайве.
- valence - показывает вероятность, с которой песня позитивная.
- time_signature - ритм песни (цифра / 4).
*Остальное по идее понятно*.

## 2) Быстрый обзор данных
### 2.1) Загрузка датасета.

In [2]:
import kagglehub
from kagglehub import KaggleDatasetAdapter
import pandas as pd

df = pd.read_csv("spotify-tracks-dataset.csv", index_col = 0) #вручную изменил название первого столбца (track_number), т.к. оно было пустое
pd.set_option('display.max_columns', None) #почему-то без этого он не выводит поле key

### 2.2) df.head(), df.tail(), df.shape.

In [3]:
display(df.head())
display(df.tail())
print("Dataset shape: ",df.shape)

,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
track_number,,,,,,,,,,,,,,,,,,,,
0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.676,0.4610,1,-6.746,0,0.1430,0.0322,0.000001,0.3580,0.715,87.917,4,acoustic
1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.420,0.1660,1,-17.235,1,0.0763,0.9240,0.000006,0.1010,0.267,77.489,4,acoustic
2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.438,0.3590,0,-9.734,1,0.0557,0.2100,0.000000,0.1170,0.120,76.332,4,acoustic
3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,Can't Help Falling In Love,71,201933,False,0.266,0.0596,0,-18.515,1,0.0363,0.9050,0.000071,0.1320,0.143,181.740,3,acoustic
4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.618,0.4430,2,-9.681,1,0.0526,0.4690,0.000000,0.0829,0.167,119.949,4,acoustic


,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
track_number,,,,,,,,,,,,,,,,,,,,
113995,2C3TZjDRiAzdyViavDJ217,Rainy Lullaby,#mindfulness - Soft Rain for Mindful Meditatio...,Sleep My Little Boy,21,384999,False,0.172,0.235,5,-16.393,1,0.0422,0.640,0.928,0.0863,0.0339,125.995,5,world-music
113996,1hIz5L4IB9hN3WRYPOCGPw,Rainy Lullaby,#mindfulness - Soft Rain for Mindful Meditatio...,Water Into Light,22,385000,False,0.174,0.117,0,-18.318,0,0.0401,0.994,0.976,0.1050,0.0350,85.239,4,world-music
113997,6x8ZfSoqDjuNa5SVP5QjvX,Cesária Evora,Best Of,Miss Perfumado,22,271466,False,0.629,0.329,0,-10.895,0,0.0420,0.867,0.000,0.0839,0.7430,132.378,4,world-music
113998,2e6sXL2bYv4bSz6VTdnfLs,Michael W. Smith,Change Your World,Friends,41,283893,False,0.587,0.506,7,-10.889,1,0.0297,0.381,0.000,0.2700,0.4130,135.960,4,world-music
113999,2hETkH7cOfqmz3LqZDHZf5,Cesária Evora,Miss Perfumado,Barbincor,22,241826,False,0.526,0.487,1,-10.204,0,0.0725,0.681,0.000,0.0893,0.7080,79.198,4,world-music


Dataset shape:  (114000, 20)


### 2.3) df.info(), df.describe(), df.describe(include = "str").

In [4]:
df.info()
display(df.describe())
display(df.describe(include = "str"))

<class 'pandas.DataFrame'>
RangeIndex: 114000 entries, 0 to 113999
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   track_id          114000 non-null  str    
 1   artists           113999 non-null  str    
 2   album_name        113999 non-null  str    
 3   track_name        113999 non-null  str    
 4   popularity        114000 non-null  int64  
 5   duration_ms       114000 non-null  int64  
 6   explicit          114000 non-null  bool   
 7   danceability      114000 non-null  float64
 8   energy            114000 non-null  float64
 9   key               114000 non-null  int64  
 10  loudness          114000 non-null  float64
 11  mode              114000 non-null  int64  
 12  speechiness       114000 non-null  float64
 13  acousticness      114000 non-null  float64
 14  instrumentalness  114000 non-null  float64
 15  liveness          114000 non-null  float64
 16  valence           114000 non-nu

,popularity,duration_ms,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature
count,114000.000000,1.140000e+05,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000
mean,33.238535,2.280292e+05,0.566800,0.641383,5.309140,-8.258960,0.637553,0.084652,0.314910,0.156050,0.213553,0.474068,122.147837,3.904035
std,22.305078,1.072977e+05,0.173542,0.251529,3.559987,5.029337,0.480709,0.105732,0.332523,0.309555,0.190378,0.259261,29.978197,0.432621
min,0.000000,0.000000e+00,0.000000,0.000000,0.000000,-49.531000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,17.000000,1.740660e+05,0.456000,0.472000,2.000000,-10.013000,0.000000,0.035900,0.016900,0.000000,0.098000,0.260000,99.218750,4.000000
50%,35.000000,2.129060e+05,0.580000,0.685000,5.000000,-7.004000,1.000000,0.048900,0.169000,0.000042,0.132000,0.464000,122.017000,4.000000
75%,50.000000,2.615060e+05,0.695000,0.854000,8.000000,-5.003000,1.000000,0.084500,0.598000,0.049000,0.273000,0.683000,140.071000,4.000000
max,100.000000,5.237295e+06,0.985000,1.000000,11.000000,4.532000,1.000000,0.965000,0.996000,1.000000,1.000000,0.995000,243.372000,5.000000


,track_id,artists,album_name,track_name,track_genre
count,114000,113999,113999,113999,114000
unique,89741,31437,46589,73608,114
top,6S3JlDAGk3uu3NtZbPnuhS,The Beatles,Alternative Christmas 2022,Run Rudolph Run,acoustic
freq,9,279,195,151,1000


### 2.4) Проверки качества: isnull().sum(), поиск дубликатов, проверка адекватности типов данных.

In [ ]:
print(df.isnull().sum())
print("Duplicates: ", df.duplicated().sum())
print(df.dtypes)

track_id            0
artists             1
album_name          1
track_name          1
popularity          0
duration_ms         0
explicit            0
danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
time_signature      0
track_genre         0
dtype: int64
Duplicated:  450
track_id                str
artists                 str
album_name              str
track_name              str
popularity            int64
duration_ms           int64
explicit               bool
danceability        float64
energy              float64
key                   int64
loudness            float64
mode                  int64
speechiness         float64
acousticness        float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
time_signature        int6

### 2.5) Мини-вывод по датасету
В датасете 114000 записей, 22 разных параметра, из них всего 450 дубликатов. Из незаполенных только 1 запись, её я просто удалю. Фактически уникально только 74к песен, которые просто относяться к разным авторам, либо совпадает название. Датасет хороший, особо чистить его как-то не надо.
> Возможно стоило конечно выбрать менее хороший датасет, но переделать будет несложно, просто другой датасет закинуть, а пока на этом сделаю.